grakel_replace/torch_wl_kernel.py

```diff
@@ -0,0 +1,172 @@
+        counts = Counter(labels.cpu().numpy())
+        feature = torch.zeros(size, device=self.device)
+        for label, count in counts.items():
+            if label < size:  # Safety check
+                feature[label] = count
+        return feature
+
+    def forward(self, graphs: list[nx.Graph]) -> torch.Tensor:
+        """Compute WL kernel matrix for a list of graphs.
+
+        Args:
+            graphs: List of NetworkX graphs
+
+        Returns:
+            Kernel matrix as a torch.Tensor
+        """
+        # Validate input
+        if (not isinstance(graphs, list) or
+            not all(isinstance(g, nx.Graph) for g in graphs)):
+            raise TypeError("Expected input type is a list of NetworkX graphs.")
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.label_dict = {}
+        self.label_counter = 0
+
+        # Handle case of empty graphs list or empty individual graphs
+        if not graphs or all(g.number_of_nodes() == 0 for g in graphs):
+            return torch.zeros((len(graphs), len(graphs)), device=self.device)
+
+        # Convert graphs to sparse adjacency matrices and initialize labels
+        adj_matrices = [self._get_sparse_adj(g) for g in graphs]
+        label_tensors = [self._init_node_labels(g) for g in graphs]
+
+        # Pre-allocate feature matrices list
+        feature_matrices = []
+
+        # First, run all iterations to compute maximum label count
+        all_label_tensors = [label_tensors]
+        for _ in range(self.n_iter):
+            new_label_tensors = [
+                self._wl_iteration(adj, labels)
+                for adj, labels in zip(adj_matrices, all_label_tensors[-1])
+            ]
+            all_label_tensors.append(new_label_tensors)
+
+        max_label_count = self.label_counter
+
+        # Now compute feature vectors for all iterations with fixed size
+        for labels_list in all_label_tensors:
+            features = torch.stack([
+                self._compute_feature_vector(labels, max_label_count)
+                for labels in labels_list
+            ])
+            feature_matrices.append(features)
+
+        # Sum up feature matrices from all iterations
+        final_features = torch.stack(feature_matrices).sum(dim=0)
+
+        # Compute kernel matrix
+        K = torch.mm(final_features, final_features.t())
+
+        # Normalize if requested
+        if self.normalize:
+            diag = torch.sqrt(torch.diag(K))
+            K = K / (diag.unsqueeze(0) * diag.unsqueeze(1))
+
+        return K
+
+
+class GraphDataset:
+    """Utility class to convert NetworkX graphs for WL kernel."""
+
+    @staticmethod
+    def from_networkx(graphs: list[nx.Graph], node_labels_tag: str = "label") -> list[
+        nx.Graph]:
+        """Convert NetworkX graphs ensuring proper node labeling."""
+        processed_graphs = []
+        for g in graphs:
+            g = g.copy()
```

Comment:
Reading the description of `nx.convert_node_labels_to_integers()`, this `copy()` isn't needed.

However this likely needs a little re-work as put in the main review.

Reply:
Thanks